From 51ffd521ffffce7736ad4d16ca7d38e2f6f00d9a Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@st.com>
Date: Fri, 1 May 2020 16:24:49 +0200
Subject: [PATCH 0280/1695] drivers: gic: restore GIC main config at resume

Split gic_init() in helper functions to have a common main configuration
load function git_init_setup() used during initialization and low power
resume sequence.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0e62ccaefeec60b4409e56929b49eb5f40f1526d drivers: gic: restore GIC main config at resume]
Change-Id: I9f13eabffc3136c6185f034e9b464287bae78bb8
Signed-off-by: Gatien Chevallier <gatien.chevallier@st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>
---
 core/drivers/gic.c | 32 ++++++++++++++++++++++----------
 1 file changed, 22 insertions(+), 10 deletions(-)

diff --git a/core/drivers/gic.c b/core/drivers/gic.c
index 367af94eb..dac031b40 100644
--- a/core/drivers/gic.c
+++ b/core/drivers/gic.c
@@ -305,13 +305,10 @@ static void gic_init_base_addr(paddr_t gicc_base_pa, paddr_t gicd_base_pa)
 	gic_pm_register(gd);
 }
 
-void gic_init(paddr_t gicc_base_pa, paddr_t gicd_base_pa)
+static void gic_setup_clear_it(struct gic_data __maybe_unused *gd)
 {
-	struct gic_data __maybe_unused *gd = &gic_data;
 	size_t __maybe_unused n = 0;
 
-	gic_init_base_addr(gicc_base_pa, gicd_base_pa);
-
 	/* GIC configuration is initialized from TF-A when embedded */
 #ifndef CFG_WITH_ARM_TRUSTED_FW
 	for (n = 0; n <= gd->max_it / NUM_INTS_PER_REG; n++) {
@@ -320,7 +317,17 @@ void gic_init(paddr_t gicc_base_pa, paddr_t gicd_base_pa)
 
 		/* Make interrupts non-pending */
 		io_write32(gd->gicd_base + GICD_ICPENDR(n), 0xffffffff);
+	}
+#endif
+}
 
+static void gic_init_setup(struct gic_data __maybe_unused *gd)
+{
+	size_t __maybe_unused n = 0;
+
+	/* GIC configuration is initialized from TF-A when embedded */
+#ifndef CFG_WITH_ARM_TRUSTED_FW
+	for (n = 0; n <= gd->max_it / NUM_INTS_PER_REG; n++) {
 		/* Mark interrupts non-secure */
 		if (n == 0) {
 			/* per-CPU inerrupts config:
@@ -351,7 +358,13 @@ void gic_init(paddr_t gicc_base_pa, paddr_t gicd_base_pa)
 		     GICD_CTLR_ENABLEGRP0 | GICD_CTLR_ENABLEGRP1);
 #endif
 #endif /*CFG_WITH_ARM_TRUSTED_FW*/
+}
 
+void gic_init(paddr_t gicc_base_pa, paddr_t gicd_base_pa)
+{
+	gic_init_base_addr(gicc_base_pa, gicd_base_pa);
+	gic_setup_clear_it(&gic_data);
+	gic_init_setup(&gic_data);
 	interrupt_main_init(&gic_data.chip);
 }
 
@@ -766,13 +779,11 @@ static void gic_save_it(struct gic_data *gd, struct gic_it_pm *pm)
 static void gic_restore_it(struct gic_data *gd, struct gic_it_pm *pm)
 {
 	unsigned int it = pm->it;
-	size_t idx = 0;
+	size_t idx = it / NUM_INTS_PER_REG;
 	uint32_t mask = BIT(it % NUM_INTS_PER_REG);
 	uint32_t shift2 = it % (NUM_INTS_PER_REG / 2) * 2;
 	uint32_t shift8 = it % (NUM_INTS_PER_REG / 8) * 8;
 
-	idx = it / NUM_INTS_PER_REG;
-
 	io_mask32(gd->gicd_base + GICD_IGROUPR(idx),
 		  (pm->flags & IT_PM_GPOUP1_BIT) ? mask : 0, mask);
 
@@ -808,11 +819,12 @@ static TEE_Result gic_pm(enum pm_op op, unsigned int pm_hint __unused,
 	struct gic_data *gd = (struct gic_data *)PM_CALLBACK_GET_HANDLE(handle);
 	struct gic_pm *pm = &gd->pm;
 
-	if (op == PM_OP_SUSPEND)
+	if (op == PM_OP_SUSPEND) {
 		sequence = gic_save_it;
-	else
+	} else {
+		gic_init_setup(gd);
 		sequence = gic_restore_it;
-
+	}
 	for (n = 0, cfg = pm->pm_cfg; n < pm->count; n++, cfg++)
 		sequence(gd, cfg);
 
-- 
2.47.2

